[HELP WANTED] Try to fix windows and external command bug

tompng opened the pull request:
This pull request tries to fix #545 and #676.
As described in #676, `@GetConsoleScreenBufferInfo.call(@hConsoleHandle, csbi)` fails after executing some external command.

> Fiddle.win32_last_error shows 6, which seems to map to ERROR_INVALID_HANDLE

I think console handle is somehow closed and invalidated by child process.
If that's true, (unconfirmed), we need to refresh `@hConsoleHandle` and retry the failed call.

# Help Wanted

This patch is not tested. I don't have an environment to reproduce.

andrewngo commented:
Seeing a load error
```
gem 'reline', git: 'https://github.com/ruby/reline.git', branch: 'windows_handler_retry'
```
and then with
```
bundle exec irb

irb(main):001:0> <gem dir>/.gem/ruby/3.2.0/bundler/gems/reline-f7f2be7db418/lib/reline/io/windows.rb:121:in `call': wrong number of arguments (given 0, expected 2) (ArgumentError)

        ret, = @func.call(*args)
                          ^^^^^
        from <gem dir>/.gem/ruby/3.2.0/bundler/gems/reline-f7f2be7db418/lib/reline/io/windows.rb:121:in `call'
        from <gem dir>/.gem/ruby/3.2.0/bundler/gems/reline-f7f2be7db418/lib/reline/io/windows.rb:465:in `show_cursor'
        from <gem dir>/.gem/ruby/3.2.0/bundler/gems/reline-f7f2be7db418/lib/reline/line_editor.rb:549:in `render_differential'
        from <gem dir>/.gem/ruby/3.2.0/bundler/gems/reline-f7f2be7db418/lib/reline/line_editor.rb:567:in `rerender'
        from <gem dir>/.gem/ruby/3.2.0/bundler/gems/reline-f7f2be7db418/lib/reline.rb:336:in `inner_readline'
        from <gem dir>/.gem/ruby/3.2.0/bundler/gems/reline-f7f2be7db418/lib/reline.rb:258:in `block (2 levels) in readmultiline'
        from <gem dir>/.gem/ruby/3.2.0/bundler/gems/reline-f7f2be7db418/lib/reline/io/windows.rb:313:in `with_raw_input'
        from <gem dir>/.gem/ruby/3.2.0/bundler/gems/reline-f7f2be7db418/lib/reline.rb:257:in `block in readmultiline'
        from <gem dir>/.gem/ruby/3.2.0/bundler/gems/reline-f7f2be7db418/lib/reline.rb:252:in `synchronize'
        from <gem dir>/.gem/ruby/3.2.0/bundler/gems/reline-f7f2be7db418/lib/reline.rb:252:in `readmultiline'
        from <ruby dir>/lib/ruby/3.2.0/forwardable.rb:240:in `readmultiline'
        from <ruby dir>/lib/ruby/3.2.0/forwardable.rb:240:in `readmultiline'
        from <ruby dir>/lib/ruby/3.2.0/irb/input-method.rb:422:in `gets'
        from <ruby dir>/lib/ruby/3.2.0/irb.rb:548:in `block (2 levels) in eval_input'
        from <ruby dir>/lib/ruby/3.2.0/irb.rb:777:in `signal_status'
        from <ruby dir>/lib/ruby/3.2.0/irb.rb:547:in `block in eval_input'
        from <ruby dir>/lib/ruby/3.2.0/irb/ruby-lex.rb:283:in `lex'
        from <ruby dir>/lib/ruby/3.2.0/irb/ruby-lex.rb:252:in `block (2 levels) in each_top_level_statement'
        from <ruby dir>/lib/ruby/3.2.0/irb/ruby-lex.rb:249:in `loop'
        from <ruby dir>/lib/ruby/3.2.0/irb/ruby-lex.rb:249:in `block in each_top_level_statement'
        from <ruby dir>/lib/ruby/3.2.0/irb/ruby-lex.rb:248:in `catch'
        from <ruby dir>/lib/ruby/3.2.0/irb/ruby-lex.rb:248:in `each_top_level_statement'
        from <ruby dir>/lib/ruby/3.2.0/irb.rb:566:in `eval_input'
        from <ruby dir>/lib/ruby/3.2.0/irb.rb:500:in `block in run'
        from <ruby dir>/lib/ruby/3.2.0/irb.rb:499:in `catch'
        from <ruby dir>/lib/ruby/3.2.0/irb.rb:499:in `run'
        from <ruby dir>/lib/ruby/3.2.0/irb.rb:421:in `start'
        from <ruby dir>/lib/ruby/gems/3.2.0/gems/irb-1.6.2/exe/irb:11:in `<top (required)>'
        from <ruby dir>/bin/irb:33:in `load'
        from <ruby dir>/bin/irb:33:in `<main>'
```

andrewngo commented:
I just tried the script in https://github.com/ruby/reline/issues/545 and can no longer reproduce the issue per the video. Seems great in my ad-hoc testing

andrewngo commented:
More ad-hoc lines
```
bundle exec irb
irb(main):001:0> Reline::IOGate.get_screen_size
=> [30, 120]
irb(main):002:0> Reline::IOGate.get_console_screen_buffer_info
=> "x\x00\x1E\x00\x00\x00\x19\x00\a\x00\x00\x00\x00\x00w\x00\x1D\x00x\x00\x1E\x00"
irb(main):003:0> STDIN.winsize
(irb):3:in `winsize': Bad file descriptor - <STDIN> (Errno::EBADF)
        from (irb):3:in `<main>'
        from <ruby dir>/lib/ruby/gems/3.2.0/gems/irb-1.6.2/exe/irb:11:in `<top (required)>'
        from <ruby dir>/bin/irb:33:in `load'
        from <ruby dir>/bin/irb:33:in `<main>'
irb(main):004:0> `ver`










Retry with refreshed console handle succeeded.










=> "\nMicrosoft Windows [Version 10.0.22631.3880]\n"
irb(main):005:0> `ruby -v`










Retry with refreshed console handle succeeded.










=> "ruby 3.2.3 (2024-01-18 revision 52bb2ac0a6) [x64-mingw-ucrt]\n"
irb(main):006:0> `chromedriver -v`










Retry with refreshed console handle succeeded.










=> "ChromeDriver 127.0.6533.99 (f31af5097d90ef5ae5bd7b8700199bc6189ba34d-refs/branch-heads/6533@{#1910})\n"
irb(main):007:0> Reline::IOGate.get_screen_size
=> [30, 120]
irb(main):008:0> Reline::IOGate.get_console_screen_buffer_info
=> "x\x00\x1E\x00\x00\x00\x1C\x00\a\x00\x00\x00\x00\x00w\x00\x1D\x00x\x00\x1E\x00"
irb(main):009:0> STDIN.winsize
(irb):9:in `winsize': Bad file descriptor - <STDIN> (Errno::EBADF)
        from (irb):9:in `<main>'
        from <ruby dir>/lib/ruby/gems/3.2.0/gems/irb-1.6.2/exe/irb:11:in `<top (required)>'
        from <ruby dir>/bin/irb:33:in `load'
        from <ruby dir>/bin/irb:33:in `<main>'
```

andrewngo-docusign commented on lib/reline/io/windows.rb:
```diff
@@ -439,14 +455,14 @@ def hide_cursor
     visible = 1 # 1 means true
     cursor_info = [size, visible].pack('Li')
-    @SetConsoleCursorInfo.call(@hConsoleHandle, cursor_info)
+    call_with_console_handle(@SetConsoleCursorInfo.call, cursor_info)
```
the `.call` is causing the error later and seems unnecessary

tompng commented:
Thank you for testing :+1:
Looks like the workaround, refresh `@hConsoleHandle` and retry, is working.


Fixed the `wrong number of arguments` bug
```diff
-    call_with_console_handle(@SetConsoleCursorInfo.call, cursor_info)
+    call_with_console_handle(@SetConsoleCursorInfo, cursor_info)
```
and removed this debug code
```diff
-    val = win32func.call(@hConsoleHandle, *args)
-
-    # Remove this after confirming this patch fixed the issue
-    10.times { puts }
-    puts "Retry with refreshed console handle #{val == 0 ? 'failed. This patch has no effect' : 'succeeded'}."
-    10.times { puts }
-    val
+    win32func.call(@hConsoleHandle, *args)
```

andrewngo commented:
Just wanted to confirm again `reline (0.5.9 5dc08c3)` looks good to me. Looking forward to this release! Thanks

hogelog reviewed (commented):
I also confirmed that the issues #545 and #676 frequently occur with Ruby 3.3.4.
But the problem no longer reproduces when using this pull request.
I don't fully understand the technical details of the changes, but it seems to be a good fix.

## Environment
Ruby version: 3.3.4
Ruby distribution: Ruby Installer 3.3.4-1 (x64) https://rubyinstaller.org/downloads/
OS: Windows 11 Home 64bit (10.0, Build 22631)
Terminal: cmd.exe, powershell.exe

st0012 reviewed (approved):
I'm good with merging this with users' confirmation. Also, great job fixing a bug that you can't test yourself 👍

ima1zumi reviewed (approved):
🚀
